src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+                {
+                    _backAddArray[i] = data[i];
+                }
+                for (int i = 0; i < _backAddWindowSize; ++i)
+                {
+                    _backAddArray[data.Length + i] = predictedValue;
+                }
+                return _backAddArray;
+            }
+
+            private double PredictNext(double[] data)
+            {
+                var n = data.Length;
+                double slopeSum = 0.0f;
+                for (int i = 0; i < n - 1; ++i)
+                {
+                    slopeSum += (data[n - 1] - data[i]) / (n - 1 - i);
+                }
+                return (data[1] + slopeSum);
+            }
+
+            private double[] AverageFilter(double[] data, int n)
+            {
+                double cumsum = 0.0f;
+                int length = data.Length;
+
+                AllocateDoubleArray(ref _cumSumList, length);
+                AllocateDoubleArray(ref _cumSumShift, length);
+
+                for (int i = 0; i < length; ++i)
+                {
+                    cumsum += data[i];
+                    _cumSumList[i] = cumsum;
+                    _cumSumShift[i] = cumsum;
+                }
+                for (int i = n; i < length; ++i)
+                {
+                    _cumSumList[i] = (_cumSumList[i] - _cumSumShift[i - n]) / n;
+                }
+                for (int i = 1; i < n; ++i)
+                {
+                    _cumSumList[i] /= (i + 1);
+                }
+                return _cumSumList;
+            }
+
+            private double CalculateScore(double mag, double avgMag)
+            {
+                double safeDivisor = avgMag;
+                if (Math.Abs(safeDivisor) < _eps)
+                {
+                    safeDivisor = _eps;
+                }
+                return (Math.Abs(mag - avgMag) / safeDivisor);
+            }
+
+            private void GetExpectedValue(double[] values, double[][] results)
+            {
+                //Step 8: Calculate Expected Value
+                var exps = CalculateExpectedValueByFft(GetDeanomalyData(values, GetAnomalyIndex(results.Select(x => x[1]).ToArray())));
+
+                for (int i = 0; i < results.Length; ++i)
+                {
+                    results[i][3] = exps[i];
+                }
+            }
+
+            private void GetMargin(double[] values, double[][] results, double sensitivity)
+            {
+                //Step 8: Calculate Expected Value
+                var exps = CalculateExpectedValueByFft(GetDeanomalyData(values, GetAnomalyIndex(results.Select(x => x[1]).ToArray())));
+
+                //Step 9: Calculate Boundary Unit
+                var units = CalculateBoundaryUnit(values, results.Select(x => x[0] > 0).ToArray());
+
+                //Step 10: Calculate UpperBound and LowerBound
+                var margins = units.Select(x => CalculateMargin(x, sensitivity)).ToList();
```

Comment:
>CalculateMargin [](start = 48, length = 15)

I would do this inside the for loop:
```
for (int i = 0; i < results.Length; ++i)
                {
                    var margin = CalculateMargin(units[i], sensitivity);
                    results[i][3] = exps[i];
                    results[i][4] = units[i];
                    results[i][5] = exps[i] + margin;
                    results[i][6] = exps[i] - margin;
                    //Step 11: Update Anomaly Score
                    results[i][1] = CalculateAnomalyScore(values[i], exps[i], units[i], results[i][0] > 0);
                }
```

Reply:
Done.

---
In reply to: [427869378](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427869378) [](ancestors = 427869378)